gobblin-data-management/src/main/java/org/apache/gobblin/data/management/copy/iceberg/IcebergPartitionDataset.java

```diff
@@ -0,0 +1,204 @@
+              actualSourceFs, srcFileStatus, targetFs.makeQualified(destPath), copyConfig)
+          .fileSet(fileSet)
+          .datasetOutputPath(targetFs.getUri().getPath())
+          .build();
+
+      fileEntity.setSourceData(getSourceDataset(this.sourceFs));
+      fileEntity.setDestinationData(getDestinationDataset(targetFs));
+      copyEntities.add(fileEntity);
+    }
+
+    // Adding this check to avoid adding post publish step when there are no files to copy.
+    if (CollectionUtils.isNotEmpty(destDataFiles)) {
+      copyEntities.add(createPostPublishStep(destDataFiles));
+    }
+
+    log.info("~{}~ generated {} copy--entities", fileSet, copyEntities.size());
+    return copyEntities;
+  }
+
+  private List<DataFile> getDestDataFiles(List<DataFile> srcDataFiles) throws IcebergTable.TableNotFoundException {
+    List<DataFile> destDataFiles = new ArrayList<>();
+    if (srcDataFiles.isEmpty()) {
+      return destDataFiles;
+    }
+    TableMetadata srcTableMetadata = getSrcIcebergTable().accessTableMetadata();
+    TableMetadata destTableMetadata = getDestIcebergTable().accessTableMetadata();
+    PartitionSpec partitionSpec = destTableMetadata.spec();
+    String srcWriteDataLocation = srcTableMetadata.property(TableProperties.WRITE_DATA_LOCATION, "");
+    String destWriteDataLocation = destTableMetadata.property(TableProperties.WRITE_DATA_LOCATION, "");
+    if (StringUtils.isEmpty(srcWriteDataLocation) || StringUtils.isEmpty(destWriteDataLocation)) {
+      log.warn(
+          String.format("Either source or destination table does not have write data location : source table write data location : {%s} , destination table write data location : {%s}",
+              srcWriteDataLocation,
+              destWriteDataLocation
+          )
+      );
+    }
+    // tableMetadata.property(TableProperties.WRITE_DATA_LOCATION, "") returns null if the property is not set and
+    // doesn't respect passed default value, so to avoid NPE in .replace() we are setting it to empty string.
+    String prefixToBeReplaced = (srcWriteDataLocation != null) ? srcWriteDataLocation : "";
+    String prefixToReplaceWith = (destWriteDataLocation != null) ? destWriteDataLocation : "";
+    srcDataFiles.forEach(dataFile -> {
+      String curDestFilePath = dataFile.path().toString();
+      String newDestFilePath = curDestFilePath.replace(prefixToBeReplaced, prefixToReplaceWith);
+      String updatedDestFilePath = addUUIDToPath(newDestFilePath);
+      destDataFiles.add(DataFiles.builder(partitionSpec)
+          .copy(dataFile)
+          .withPath(updatedDestFilePath)
+          .build());
+    });
+    return destDataFiles;
+  }
+
+  private String addUUIDToPath(String filePathStr) {
+    Path filePath = new Path(filePathStr);
+    String fileDir = filePath.getParent().toString();
+    String fileName = filePath.getName();
+    String newFileName = UUID.randomUUID() + "-" + fileName;
+    return String.join("/", fileDir, newFileName);
```

Comment:
usually best to avoid hard-coding path separators, like `/`:
```
Path filePath = new Path(filePathStr);
String fileName = filePath.getName();
String newFileName = UUID.randomUUID() + "-" + fileName;
return new Path(filePath.getParent(), newFileName).toString()
```